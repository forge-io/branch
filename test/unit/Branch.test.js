const mod = require('../../');
const { Branch } = mod;

describe('the tree is build correctly', () => {
  let instance;
  beforeEach(() => {
    instance = new Branch();
  });

  it('for leaf regexp', () => {
    instance.add('/{param}/path', 'GET', () => {});
    expect(instance.tree).toMatchSnapshot();
  });

  it('for leaf string', () => {
    instance.add('/simplestring/path', 'GET', () => {});
    expect(instance.tree).toMatchSnapshot();
  });
});

describe('lookup returns correctly', () => {
  let instance;

  const pathA = '/handler/stringa';
  const pathAbad = '/nohandler/stringa';
  const pathAtrailingSlash = '/handler/stringa/';
  const pathB = '/handler/basdf';
  const pathC = '/handler/casdfasdf/stringc';
  const pathD = '/handler/cacacaccc/stringd';
  const pathE = '/handler/get/params?bob=sue&cat=alfie';
  const pathF = '/handler/dasdfasdfpartfffffffsextra/get';

  const expectedParamsD = {
    url: {
      paramc: 'acacaccc'
    },
    get: {}
  };

  const expectedParamsE = {
    get: {
      bob: 'sue',
      cat: 'alfie'
    },
    url: {}
  };

  const expectedParamsF = {
    url: {
      paramA: 'asdfasdf',
      paramB: 'fffffffs'
    },
    get: {}
  };

  const handlerA = () => {};
  const handlerB = () => {};
  const handlerC = () => {};
  const handlerD = () => {};
  const handlerE = () => {};
  const handlerF = () => {};

  const addTestMethods = (instance) => {
    instance.add(pathA, 'GET', handlerA);
    instance.add('/handler/b{paramb}', 'GET', handlerB);
    instance.add('/handler/c{paramc}/stringc', 'GET', handlerC);
    instance.add('/handler/c{paramc}/stringd/', 'ANY', handlerD);
    instance.add('/handler/get/params', 'ANY', handlerE);
    instance.add('/handler/d{paramA}part{paramB}extra/get', 'GET', handlerF);
    return instance;
  };

  beforeEach(() => {
    instance = new Branch();
    instance = addTestMethods(instance);
  });

  it('for simple string lookups that don\'t match', () => {
    const lookupResult = instance.lookup(pathAbad, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(false);
    expect(lookupResult.handler).toEqual(undefined);
  });

  it('for simple string lookups', () => {
    const lookupResult = instance.lookup(pathA, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerA);
  });

  it('for simple string lookups removing the trailing slash', () => {
    const lookupResult = instance.lookup(pathAtrailingSlash, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerA);
  });

  it('for mixed regexp and string lookups', () => {
    const lookupResult = instance.lookup(pathB, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerB);
  });

  it('not found when leaf doesn\'t exist', () => {
    const lookupResult = instance.lookup(pathC, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerC);
  });

  it('handles different methods falling back to ANY', () => {
    const lookupResult = instance.lookup(pathD, 'BOB');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerD);
    expect(lookupResult.params).toEqual(expectedParamsD);
  });

  it('handles different methods falling back to ANY', () => {
    const lookupResult = instance.lookup(pathE, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerE);
    expect(lookupResult.params).toEqual(expectedParamsE);
  });

  it('handles regexp extra bits on params', () => {
    const lookupResult = instance.lookup(pathF, 'GET');
    expect(lookupResult).toMatchSnapshot();
    expect(lookupResult.foundRoute).toBe(true);
    expect(lookupResult.handler).toEqual(handlerF);
    expect(lookupResult.params).toEqual(expectedParamsF);
  });
});
