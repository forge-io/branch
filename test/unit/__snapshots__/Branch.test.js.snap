// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lookup returns correctly for mixed regexp and string lookups 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {
      "paramb": "asdf",
    },
  },
}
`;

exports[`lookup returns correctly for simple string lookups 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {},
  },
}
`;

exports[`lookup returns correctly for simple string lookups removing the trailing slash 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {},
  },
}
`;

exports[`lookup returns correctly for simple string lookups that don't match 1`] = `
Object {
  "foundRoute": false,
  "handler": undefined,
  "params": Object {
    "get": Object {},
    "url": Object {},
  },
}
`;

exports[`lookup returns correctly handles different methods falling back to ANY 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {
      "paramc": "acacaccc",
    },
  },
}
`;

exports[`lookup returns correctly handles different methods falling back to ANY 2`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {
      "bob": "sue",
      "cat": "alfie",
    },
    "url": Object {},
  },
}
`;

exports[`lookup returns correctly handles regexp extra bits on params 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {
      "paramA": "asdfasdf",
      "paramB": "fffffffs",
    },
  },
}
`;

exports[`lookup returns correctly not found when leaf doesn't exist 1`] = `
Object {
  "foundRoute": true,
  "handler": [Function],
  "params": Object {
    "get": Object {},
    "url": Object {
      "paramc": "asdfasdf",
    },
  },
}
`;

exports[`the tree is build correctly for leaf regexp 1`] = `
Leaf {
  "children": Object {
    "regexps": Object {
      "{param}": Leaf {
        "children": Object {
          "regexps": Object {},
          "strings": Object {
            "path": Leaf {
              "children": Object {
                "regexps": Object {},
                "strings": Object {},
              },
              "handlers": Object {
                "GET": [Function],
              },
              "regexp": null,
            },
          },
        },
        "handlers": Object {},
        "regexp": /\\(\\?<param>\\.\\*\\)/,
      },
    },
    "strings": Object {},
  },
  "handlers": Object {},
  "regexp": null,
}
`;

exports[`the tree is build correctly for leaf string 1`] = `
Leaf {
  "children": Object {
    "regexps": Object {},
    "strings": Object {
      "simplestring": Leaf {
        "children": Object {
          "regexps": Object {},
          "strings": Object {
            "path": Leaf {
              "children": Object {
                "regexps": Object {},
                "strings": Object {},
              },
              "handlers": Object {
                "GET": [Function],
              },
              "regexp": null,
            },
          },
        },
        "handlers": Object {},
        "regexp": null,
      },
    },
  },
  "handlers": Object {},
  "regexp": null,
}
`;
